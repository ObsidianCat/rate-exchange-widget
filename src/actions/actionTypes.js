export const FETCH_RATES = "FETCH_RATES";
export const SET_ORIGIN_CURRENCY_TYPE = "SET_ORIGIN_CURRENCY_TYPE";
export const SET_TARGET_CURRENCY_TYPE = "SET_TARGET_CURRENCY_TYPE";
export const SET_AMOUNT = "SET_AMOUNT";
export const SWAP_CURRENCY_TYPE = "SWAP_CURRENCY_TYPE";
