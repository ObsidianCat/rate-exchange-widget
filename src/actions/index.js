export * from './actionTypes'
export * from './exchangeActions'
export * from './ratesActions'