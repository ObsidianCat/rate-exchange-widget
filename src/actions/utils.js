import axios from "axios";

export const EXCHANGE_API_BASE = "https://api.exchangeratesapi.io/latest"

export function getRates(currency){
  return axios.get(`${EXCHANGE_API_BASE}?base=${currency}`)
}