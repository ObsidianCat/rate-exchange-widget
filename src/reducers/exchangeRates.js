import { FETCH_RATES } from "../actions/actionTypes";

export default function exchangeRates(state = {}, action) {
  if (action.type === FETCH_RATES) {
    return action.payload;
  } else {
    return state;
  }
}
