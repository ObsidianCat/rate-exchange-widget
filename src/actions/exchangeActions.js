import axios from "axios";
import * as CONSTANTS from './actionTypes'
import {SET_ORIGIN_CURRENCY_TYPE, SET_TARGET_CURRENCY_TYPE} from "./actionTypes";
import {SWAP_CURRENCY_TYPE} from "./actionTypes";
import {calculateTargetFromOrigin} from '../utils'
import {SET_AMOUNT} from "./actionTypes";
import {getRates} from "./utils";


export function setCurrencyTypeOrigin(currencyType) {
  return async function setCurrencyTypeOriginThunk(dispatch) {
    try {
      const response = await  getRates(currencyType)
      dispatch({ type: CONSTANTS.FETCH_RATES, payload: response.data.rates  })
      dispatch({ type: SET_ORIGIN_CURRENCY_TYPE, payload: currencyType  })
    } catch(error){
      console.error(error)
    }
  }
}

export function setCurrencyTypeTarget(payload) {
  return { type: SET_TARGET_CURRENCY_TYPE, payload: payload }
}

export function updateOriginAmount(payload){
  return { type: CONSTANTS.SET_AMOUNT, payload }
}

export function swapCurrencyType(){
  return { type: SWAP_CURRENCY_TYPE }
}

export function doSwap(){
  return async function doSwapThunk(dispatch, getState){
    const { exchange, exchangeRates } = getState()
    const futureAmountOrigin = calculateTargetFromOrigin(exchange.amount, exchangeRates, exchange.targetCurrency)
    dispatch({type: CONSTANTS.SWAP_CURRENCY_TYPE})
    const response = await  getRates(getState().exchange.originCurrency)
    dispatch({ type: CONSTANTS.FETCH_RATES, payload: response.data.rates  })
    dispatch({type: SET_AMOUNT, payload: futureAmountOrigin})

  }
}