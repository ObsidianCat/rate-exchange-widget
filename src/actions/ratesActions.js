import * as CONSTANTS from "./actionTypes";
import {getRates} from "./utils";


export function fetchExchangeRates() {
  return async function fetchExchangeRateThunk(dispatch, getState) {
    try {
      const { exchange } = getState()
      const response = await  getRates(exchange.originCurrency)
      dispatch({ type: CONSTANTS.FETCH_RATES, payload: response.data.rates  })
    } catch(error){
      console.error(error)
    }
  }
}