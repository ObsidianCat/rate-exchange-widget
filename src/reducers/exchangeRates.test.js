import { FETCH_RATES } from "../actions/actionTypes";
import exchangeRates from "./exchangeRates"

//Example of how I would approach testing of reducers. The second reducer of this app can be tested in similar fashion
describe('exchangeRates reducer', () => {
  it('should return the initial state', () => {
    expect(exchangeRates(undefined, {})).toMatchSnapshot()
  })

  it('should populate state with received data', () => {
    const mockFetchRatesAction = {
      type: FETCH_RATES,
      payload: {
        "USD": 321.97,
        "CAD": 1.5641,
        }
    }
    expect(exchangeRates({},mockFetchRatesAction)).toMatchSnapshot()
  })
})