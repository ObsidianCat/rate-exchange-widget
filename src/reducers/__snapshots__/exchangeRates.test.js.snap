// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exchangeRates reducer should populate state with received data 1`] = `
Object {
  "CAD": 1.5641,
  "USD": 321.97,
}
`;

exports[`exchangeRates reducer should return the initial state 1`] = `Object {}`;
